### DISCUSSION TITLE
Fit bounds with padding

### DISCUSSION THREAD

**Discussion by ariel-upstream** [visgl/deck.gl#7526] [Q&A] [answered]
Hi, I'm trying to use fiBounds with the WebMercatorViewport object but getting an error about the padding property.
code: 
```
 const initialViewState: IViewport | WebMercatorViewport = useMemo(() => {
    if (locationsFitBounds) {
      const minLat: number = Math.min(...locationsFitBounds.map((l: MapLocation) => l.latitude))
      const maxLat: number = Math.max(...locationsFitBounds.map((l: MapLocation) => l.latitude))
      const minLng: number = Math.min(...locationsFitBounds.map((l: MapLocation) => l.longitude))
      const maxLng: number = Math.max(...locationsFitBounds.map((l: MapLocation) => l.longitude))
      const southWest: [number, number] = [minLng, minLat]
      const northEast: [number, number] = [maxLng, maxLat]
        return new WebMercatorViewport({ width: 800, height: 600 }).fitBounds([southWest, northEast], { padding: 150 })
    } else {
      return viewport
    }
  }, [locationsFitBounds, viewport])
```

the inital render is working good but when I'm trying to close the map I'm getting the error: 
`react-dom.development.js:26874 Uncaught TypeError: Cannot assign to read only property 'padding' of object '#<WebMercatorViewport>'
    at positionChildrenUnderViews (position-children-under-views.ts:60:1)
    at deckgl.ts:254:1
    at renderWithHooks (react-dom.development.js:16175:1)
    at updateForwardRef (react-dom.development.js:20023:1)
    at beginWork (react-dom.development.js:22465:1)
    at beginWork$1 (react-dom.development.js:27381:1)
    at performUnitOfWork (react-dom.development.js:26513:1)
    at workLoopSync (react-dom.development.js:26422:1)
    at renderRootSync (react-dom.development.js:26390:1)
    at recoverFromConcurrentError (react-dom.development.js:25806:1)`

```
<React.Fragment>
  <DeckGL
    width="100%"
    height="100%"
    controller={true}
    initialViewState={initialViewState}
  >
    <Map
      attributionControl={false}
      style={{ width: "100%", height: "100%" }}
    />
  </DeckGL>
</React.Fragment>
```

I'm using: deck.gl: "8.8.20" react-map-gl": "7.0.20"

**Comment by Pessimistress** [comment, marked as answer]
You can't pass a WebMercatorViewport instance to `viewState`/`initialViewState` because, as the error suggests, it are immutable. Please consult the documentation of these props.

**Reply by ariel-upstream** [reply]
Can you elaborate on how should I use the WebMercatorViewport? couldn't find a deck.gl wrapping react-map-gl example with fitBounds

**Reply by Pessimistress** [reply]
```js
const initialViewState = useMemo(() => {
    ...
    const {longitude, latitude, zoom} = new WebMercatorViewport({ width: 800, height: 600 }).fitBounds([southWest, northEast])
    return {longitude, latitude, zoom}
  }, [locationsFitBounds])
```

**Reply by ilyabo** [reply]
I stumbled upon this issue too when trying to save the map viewport in a global app state using a library  which freezes the objects before saving them to ensure immutability ([immer](https://immerjs.github.io/immer/)). Maybe `positionChildrenUnderViews()` could create a copy of the viewport when setting the padding to avoid the error?

**Reply by pequeurb** [reply]
I had a similar issue, but was able to fix it thanks to the comment above.
I upgraded deck.gl to the latest version and started getting this issue as a result, because I keep the initial view state in redux/immer.
For now I'm working around it, but I agree with the comment above that it would probably be a good idea to make a copy instead of directly mutating the state that is passed.
